toolchain/parse/handle_import_and_package.cpp

```diff
@@ -157,49 +166,106 @@ static auto HandleDirectiveContent(Context& context,
     case Context::PackagingState::FileStart:
       // `package` is no longer allowed, but `import` may repeat.
       context.set_packaging_state(Context::PackagingState::InImports);
-      [[fallthrough]];
+      return true;
 
     case Context::PackagingState::InImports:
-      HandleDirectiveContent(context, state, directive, on_parse_error);
-      break;
+      return true;
 
     case Context::PackagingState::AfterNonPackagingDecl: {
       context.set_packaging_state(
           Context::PackagingState::InImportsAfterNonPackagingDecl);
-      CARBON_DIAGNOSTIC(
-          ImportTooLate, Error,
-          "`import` directives must come after the `package` directive (if "
-          "present) and before any other entities in the file.");
+      CARBON_DIAGNOSTIC(ImportTooLate, Error,
+                        "`import` and `export` directives must come after the "
+                        "`package` directive (if present) and before any other "
+                        "entities in the file.");
       CARBON_DIAGNOSTIC(FirstDecl, Note, "First declaration is here.");
       context.emitter()
           .Build(intro_token, ImportTooLate)
           .Note(context.first_non_packaging_token(), FirstDecl)
           .Emit();
-      on_parse_error();
-      break;
+      return false;
     }
+
     case Context::PackagingState::InImportsAfterNonPackagingDecl:
       // There is a sequential block of misplaced `import` statements, which can
       // occur if a declaration is added above `import`s. Avoid duplicate
       // warnings.
-      on_parse_error();
-      break;
+      return false;
   }
 }
 
+// Common logic for both `import` and `export import`, distinguished by whether
+// `export_token` is valid.
+static auto HandleImportHelper(Context& context,
+                               const Context::StateStackEntry& state,
+                               Lex::TokenIndex export_token) -> void {
+  auto directive = NodeKind::ImportDirective;
+  auto on_parse_error = [&] { OnParseError(context, state, directive); };
+
+  auto intro_token = context.ConsumeChecked(Lex::TokenKind::Import);
+  context.AddLeafNode(NodeKind::ImportIntroducer, intro_token);
+
+  if (export_token.is_valid()) {
+    context.AddLeafNode(NodeKind::ImportExport, export_token);
+  }
+
+  if (VerifyInImports(context, intro_token)) {
+    HandleDirectiveContent(context, state, directive, export_token.is_valid(),
+                           on_parse_error);
+  } else {
+    on_parse_error();
+  }
+}
+
+auto HandleExport(Context& context) -> void {
+  auto state = context.PopState();
+
+  context.ConsumeChecked(Lex::TokenKind::Export);
+
+  // Error for both Main//default and every implementation file.
+  auto packaging = context.tree().packaging_directive();
+  if (!packaging || packaging->api_or_impl == Tree::ApiOrImpl::Impl) {
+    CARBON_DIAGNOSTIC(ExportFromImpl, Error,
+                      "`export` is only allowed in API files.");
```

Comment:
Given that we decided to consider the default file to be an API file rather than an implementation file, this error seems to be technically incorrect in that case. Is it worth adding a custom diagnostic for that?

Reply:
Let's address any decision from #naming in a separate PR.